jupyterhub_traefik_proxy/etcd.py

```diff
@@ -0,0 +1,271 @@
+"""Traefik implementation
+
+Custom proxy implementations can subclass :class:`Proxy`
+and register in JupyterHub config:
+
+.. sourcecode:: python
+
+    from mymodule import MyProxy
+    c.JupyterHub.proxy_class = MyProxy
+
+Route Specification:
+
+- A routespec is a URL prefix ([host]/path/), e.g.
+  'host.tld/path/' for host-based routing or '/path/' for default routing.
+- Route paths should be normalized to always start and end with '/'
+"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import asyncio
+import json
+import etcd3
+import os
+
+from subprocess import Popen
+
+from traitlets import Any, Unicode
+
+from jupyterhub.proxy import Proxy
+
+from . import traefik_utils
+
+
+class TraefikEtcdProxy(Proxy):
+    """JupyterHub Proxy implementation using traefik and etcd"""
+
+    traefik_process = Any()
+    client = etcd3.client("127.0.0.1", 2379)
```

Comment:
We probably don't want too class-level atributes of clients like this. Instead, maybe:

```python
from traitlets import default
...
client = Any()
@default('client')
def _default_client(self):
    return etc3.client(...)
```